app/controllers/ConfirmSubmissionController.scala

```diff
@@ -140,4 +147,68 @@ class ConfirmSubmissionController @Inject()(val messagesApi: MessagesApi,
       Ok
     }
   }
+
+  private val authRetrievals = Retrievals.affinityGroup and Retrievals.internalId and
+    Retrievals.externalId and Retrievals.agentCode and
+    Retrievals.credentials and Retrievals.confidenceLevel and
+    Retrievals.nino and Retrievals.saUtr and
+    Retrievals.name and Retrievals.dateOfBirth and
+    Retrievals.email and Retrievals.agentInformation and
+    Retrievals.groupIdentifier and Retrievals.credentialRole and
+    Retrievals.mdtpInformation and Retrievals.itmpName and
+    Retrievals.itmpDateOfBirth and Retrievals.itmpAddress and
+    Retrievals.credentialStrength and Retrievals.loginTimes
+
+  private[controllers] def buildIdentityData[A]()(implicit user: User[A]): Future[Either[Result, IdentityData]] = {
+    authService.authorised().retrieve(authRetrievals) {
+
+      case affinityGroup ~ internalId ~
+        externalId ~ agentCode ~
+        Some(credentials) ~ confidenceLevel ~
+        nino ~ saUtr ~
+        Some(name) ~ dateOfBirth ~
+        email ~ agentInfo ~
+        groupId ~ credentialRole ~
+        mdtpInfo ~ itmpName ~
+        itmpDateOfBirth ~ itmpAddress ~
+        credentialStrength ~ loginTimes =>
+
+        val itmpData = handleITMPData(itmpName, itmpAddress)
+        val identityData = IdentityData(internalId, externalId, agentCode,
+          credentials, confidenceLevel, nino, saUtr, name, dateOfBirth,
+          email, agentInfo, groupIdentifier = groupId,
+          credentialRole, mdtpInformation = mdtpInfo,
+          itmpName = itmpData._1, itmpDateOfBirth,
+          itmpAddress = itmpData._2, affinityGroup, credentialStrength,
+          loginTimes = IdentityLoginTimes(
+            LocalDateTime.ofInstant(Instant.parse(loginTimes.currentLogin.toInstant.toString), ZoneId.of("UTC")),
+            loginTimes.previousLogin.map(dateTime => LocalDateTime.ofInstant(Instant.parse(dateTime.toInstant.toString), ZoneId.of("UTC")))
+          )
+        )
+
+        Future.successful(Right(identityData))
+
+      case _ =>
+        Logger.warn("[ConfirmSubmission][buildIdentityData] - Did not receive minimum data from Auth required for NRS Submission")
+        Future.successful(Left(errorHandler.showInternalServerError))
+    }
+  } recover {
+    case exception: AuthorisationException =>
+      Logger.warn(s"[ConfirmSubmissionController][buildIdentityData]" +
+        s" Client authorisation failed due to internal server error. auth-client exception was $exception")
+      Left(errorHandler.showInternalServerError)
+  }
+
+  private[controllers] def handleITMPData(itmpName: Option[ItmpName], itmpAddress: Option[ItmpAddress]): (ItmpName, ItmpAddress) = {
```

Comment:
You could simplify this a lot and avoid returning tuples by having two short methods:
```
private[controllers] def handleITMPName(itmpName: Option[ItmpName]): ItmpName = itmpName.getOrElse(ItmpName(None, None, None))
private[controllers] def handleITMPAddress(itmpAddress: Option[ItmpAddress]): ItmpAddress = itmpAddress.getOrElse(ItmpAddress(None, None, None, None, None, None, None, None))
```
Then in buildIdentityData call the relevant method to handle the optional values returned by auth retrievals.